File: chtholly/src/parser/parser.rs
use super::{ast::*, lexer::*};

pub struct Parser<'a> {
    lexer: Lexer<'a>,
    current_token: Token,
}

impl<'a> Parser<'a> {
    pub fn new(lexer: Lexer<'a>) -> Self {
        let mut parser = Parser {
            lexer,
            current_token: Token::Eof,
        };
        // Prime the parser with the first token.
        parser.next_token();
        parser
    }

    fn next_token(&mut self) {
        self.current_token = self.lexer.next_token();
    }

    pub fn parse_program(&mut self) -> Program {
        Program { nodes: vec![] }
    }
}

File: src/Parser/Parser.cpp
#include "Parser.h"
#include "../AST/StmtAST.h"
#include "../AST/ExprAST.h"
#include <stdexcept>

Parser::Parser(Lexer& lexer) : lexer(lexer) {
    currentToken = this->lexer.nextToken();
}

void Parser::eat(TokenType type) {
    if (currentToken.type == type) {
        currentToken = lexer.nextToken();
    } else {
        throw std::runtime_error("Unexpected token");
    }
}

std::unique_ptr<StmtAST> Parser::parseLetStatement() {
    eat(TokenType::Let);
    std::string variableName = currentToken.value;
    eat(TokenType::Identifier);
    eat(TokenType::Assign);
    int value = std::stoi(currentToken.value);
    eat(TokenType::Integer);
    eat(TokenType::Semicolon);

    auto numberExpr = std::make_unique<NumberExprAST>(value);
    return std::make_unique<LetStmtAST>(variableName, std::move(numberExpr));
}
